godot-core/src/builtin/rid.rs

```diff
@@ -0,0 +1,140 @@
+/// find anything suspicious or outright UB please open an issue.
+///
+/// [servers]: https://docs.godotengine.org/en/stable/tutorials/optimization/using_servers.html
+/// [docs]: https://docs.godotengine.org/en/stable/classes/class_rid.html
+
+// Using normal rust repr to take advantage advantage of the nullable pointer optimization. As this enum is
+// eligible for it, it is also guaranteed to have it. Meaning the layout of this type is identical to `u64`.
+// See: https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization
+// Cannot use `#[repr(C)]` as it does not use the nullable pointer optimization.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, Debug)]
+pub enum Rid {
+    /// A valid RID may refer to some resource, but is not guaranteed to do so.
+    Valid(NonZeroU64),
+    /// An invalid RID will never refer to a resource. Internally it is represented as a 0.
+    Invalid,
+}
+
+impl Rid {
+    /// Create a new RID.
+    #[inline]
+    pub const fn new(id: u64) -> Self {
+        match NonZeroU64::new(id) {
+            Some(id) => Self::Valid(id),
+            None => Self::Invalid,
+        }
+    }
+
+    /// Convert this RID into a [`u64`].
+    ///
+    /// _Godot equivalent: `Rid.get_id()`_
+    #[inline]
+    pub const fn to_u64(self) -> u64 {
+        match self {
+            Rid::Valid(id) => id.get(),
+            Rid::Invalid => 0,
+        }
+    }
+
+    /// Convert this RID into a [`NonZeroU64`].
+    #[inline]
+    pub const fn to_non_zero_u64(self) -> Option<NonZeroU64> {
+        match self {
+            Rid::Valid(id) => Some(id),
+            Rid::Invalid => None,
+        }
+    }
+
+    /// Returns `true` if this is a valid RID.
+    #[inline]
+    pub const fn is_valid(&self) -> bool {
+        matches!(self, Rid::Valid(_))
+    }
+
+    /// Returns `true` if this is an invalid RID.
+    #[inline]
+    pub const fn is_invalid(&self) -> bool {
+        matches!(self, Rid::Invalid)
+    }
+}
+
+impl From<u64> for Rid {
+    #[inline]
+    fn from(id: u64) -> Self {
+        Self::new(id)
+    }
+}
+
+impl From<NonZeroU64> for Rid {
+    #[inline]
+    fn from(id: NonZeroU64) -> Self {
+        Self::Valid(id)
+    }
+}
+
+impl From<Rid> for u64 {
+    #[inline]
+    fn from(rid: Rid) -> Self {
+        rid.to_u64()
+    }
+}
+
+impl TryFrom<Rid> for NonZeroU64 {
+    type Error = ();
+
+    #[inline]
+    fn try_from(rid: Rid) -> Result<Self, Self::Error> {
+        rid.to_non_zero_u64().ok_or(())
+    }
+}
+
+impl From<Rid> for Option<NonZeroU64> {
+    #[inline]
+    fn from(rid: Rid) -> Self {
+        match rid {
+            Rid::Invalid => None,
+            Rid::Valid(id) => Some(id),
+        }
+    }
+}
+
+impl From<Option<NonZeroU64>> for Rid {
+    #[inline]
+    fn from(id: Option<NonZeroU64>) -> Self {
+        match id {
+            Some(id) => Rid::Valid(id),
+            None => Rid::Invalid,
+        }
+    }
+}
```

Comment:
Again, those are redundant with the inherent methods, and I'd prefer to have explicit versions rather than `323.into()`. It's also not very common that a user is actually interested in the integer itself. `Rid` is mostly intended to be an opaque ID, and there are already named conversions for the other cases.